developer_guides/virtualisation/virtualisation.rst

```diff
@@ -0,0 +1,349 @@
+
+There are different machine-level virtualisation designs, implementing different
+software configurations. Typically they will have one instance of a general
+purpose OS, running either directly on the hardware, or at least with direct
+unrestricted access to the hardware. This instance is usually called a "host."
+Within that host OS, there is usually one or more user-space processes,
+emulating further independent virtual machines, running their own complete
+instances of general-purpose OSes. Those instances are called "guests."
+
+When Linux is running in a VM, multiple drivers are already available, using
+virtio: networking, storage, etc. Those drivers use the virtio and virtqueue
+APIs to communicate with the host or, potentially, with other guests.
+
+The host part depends on the used virtualisation solution. Some of them
+implement the VirtIO API themselves, then using their method of choice to talk
+to the host they are running on. A typical Linux virtualisation implementation
+uses QEMU + KVM to run guest instances. In this case there are two ways to
+implement a VirtIO host: in the user space and in the kernel. If implemented in
+the user space, it is managed by QEMU. In the kernel space VirtIO host drivers
+use the Linux vhost API.
+
+It is also worth to note, that VirtIO guest drivers reside in respective driver
+subdirectories in the Linux kernel, e.g. drivers/net, drivers/block etc. On the
+contrary, all vhost drivers are located under drivers/vhost. That directory also
+contains vhost.c and vhost.h which are used by all vhost drivers.
+
+This document describes the Linux driver architecture to virtualise SOF DSP
+firmware features across multiple guest VMs.
+
+It describes a 
+
+1. SOF VirtIO (Virtual Frontend) driver to be run on a guest virtual machine.
+2. SOF vhost (Virtual Backend) driver to be run on the host.
+3. SOF audio topology for both host and guest.
+4. SOF supporting code in QEMU.
+
+This driver architecture allows full SOF audio use cases to reside on either
+guest VM or host. It also allows non audio DSP firmware features to also be
+virtualised on guests.
+
+Every VirtIO driver requires explicit support in QEMU, which can be very minimal
+and simple in cases where all the functionality is implemented in VirtIO and
+vhost drivers, or it can actually implement a user-space counterpart to VirtIO
+drivers on the guest. In our case the SOF QEMU support code is rather minimal,
+it only adds support for guest reset reporting and guest identigication to a set
+of standard and mostly empty device node and virtual queue management
+operations.
+
+2. SOF VirtIO design
+********************
+
+The SOF_ VirtIO design virtualises the IPC (Inter-Process Communication) command
+and audio data flows between host and guests. It re-uses the SOF-defined IPC
+API, also used by the native SOF driver to communicate with the SOF firmware,
+running on the audio DSP. This means, that the virtualisation is "applied"
```

Comment:
well, isn't it a rather standard notion for virtualisation? You can "apply" virtualisation at different levels: you can virtualise hardware, you can virtualise an OS, you can virtualise library APIs...

Reply:
re-reading that text I found the first sentence rather unclear, so I improved it. But this part I actually find rather logical: first a sentence describes, that in the native case the driver uses IPC to talk to the DSP, then the following sentence says, that the virtualisation is applied right there - *between* the driver and the DSP. But maybe it's the first sentence of the section that made that part confusing. Hopefully after I improve it, it will make the following text clearer too.